service/nodes.go

```diff
@@ -0,0 +1,121 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// nolint:staticcheck TODO reenabled staticcheck
+package service // import "go.opentelemetry.io/collector/service"
+
+import (
+	"hash/fnv"
+
+	"gonum.org/v1/gonum/graph"
+
+	"go.opentelemetry.io/collector/component"
+)
+
+type componentNodeID int64
+
+func (n componentNodeID) ID() int64 {
+	return int64(n)
+}
+
+func newComponentNodeID(parts ...string) componentNodeID {
+	h := fnv.New64a()
+	for _, part := range parts {
+		_, _ = h.Write([]byte(part))
+	}
```

Comment:
```suggestion
	h.Write([]byte(strings.Join(parts, "|"))
```

A delimiter is needed between parts to avoid accidental collision between `[]string{"ab"}` and `[]string{"a", "b"}`.